src/ray/raylet/node_manager.cc

```diff
@@ -175,31 +268,39 @@ void NodeManager::ProcessClientMessage(std::shared_ptr<LocalClientConnection> cl
     }
   } break;
   case protocol::MessageType_GetTask: {
-    const std::shared_ptr<Worker> worker = worker_pool_.GetRegisteredWorker(client);
+    std::shared_ptr<Worker> worker = worker_pool_.GetRegisteredWorker(client);
     RAY_CHECK(worker);
     // If the worker was assigned a task, mark it as finished.
     if (!worker->GetAssignedTaskId().is_nil()) {
-      FinishTask(worker->GetAssignedTaskId());
+      FinishAssignedTask(worker);
     }
     // Return the worker to the idle pool.
     worker_pool_.PushWorker(worker);
+    // Check if there is a scheduled task that can now be assigned to the newly
+    // idle worker.
     auto scheduled_tasks = local_queues_.GetScheduledTasks();
     if (!scheduled_tasks.empty()) {
-      const TaskID &scheduled_task_id =
-          scheduled_tasks.front().GetTaskSpecification().TaskId();
-      auto scheduled_tasks = local_queues_.RemoveTasks({scheduled_task_id});
-      AssignTask(scheduled_tasks.front());
+      // Find a scheduled task that whose actor ID matches that of the newly
+      // idle worker.
+      auto worker_actor_id = worker->GetActorId();
+      for (const auto &task : scheduled_tasks) {
+        if (task.GetTaskSpecification().ActorId() == worker_actor_id) {
+          auto scheduled_tasks =
+              local_queues_.RemoveTasks({task.GetTaskSpecification().TaskId()});
+          AssignTask(scheduled_tasks.front());
+        }
+      }
     }
   } break;
   case protocol::MessageType_DisconnectClient: {
     // Remove the dead worker from the pool and stop listening for messages.
     const std::shared_ptr<Worker> worker = worker_pool_.GetRegisteredWorker(client);
     if (worker) {
-      if (!worker->GetAssignedTaskId().is_nil()) {
-        // TODO(swang): Clean up any tasks that were assigned to the worker.
-        // Release any resources that may be held by this worker.
-        FinishTask(worker->GetAssignedTaskId());
-      }
+      // TODO(swang): Handle the case where the worker is killed while
+      // executing a task. Clean up the assigned task's resources, return an
+      // error to the driver.
+      // RAY_CHECK(worker->GetAssignedTaskId().is_nil())
+      //    << "Worker died while executing task: " << worker->GetAssignedTaskId();
```

Comment:
What changed here?

Reply:
This never happened before, but now it happens if the actor goes out of scope and executes the `__ray_terminate__` task, which exits mid-task. I thought it might be a good idea to try and handle this case better in the new version, so I held off on it for now.